src/main/java/org/reaktivity/nukleus/tls/internal/stream/TlsServerFactory.java

```diff
@@ -412,66 +404,71 @@ private int decodeClientHello(
         int progress,
         int limit)
     {
-        TlsRecordInfoFW tlsRecordInfo = tlsRecordInfoRO.tryWrap(buffer, progress, limit);
+        final TlsStoreInfo newTlsStoreInfo = lookupStore.apply(server.correlationRouteId);
+        final SSLContext sslContext = newTlsStoreInfo == null ? null : newTlsStoreInfo.context;
 
-        if (tlsRecordInfo != null)
+        if (sslContext != null)
         {
-            final TlsProtocolVersionFW version = tlsRecordInfo.version();
-            if (version.major() > 3)
+            final SSLEngine newTlsEngine = sslContext.createSSLEngine();
+            newTlsEngine.setUseClientMode(false);
+            if (newTlsStoreInfo.supportsClientAuth)
             {
-                tlsRecordInfo = null;
-                server.cleanupNetwork(traceId);
-                server.decoder = decodeIgnoreAll;
+                newTlsEngine.setWantClientAuth(true);
             }
-        }
 
-        if (tlsRecordInfo != null)
-        {
-            final int tlsFragmentOffset = tlsRecordInfo.limit();
-            final int tlsRecordLimit = tlsFragmentOffset + tlsRecordInfo.length();
-
-            if (limit >= tlsRecordLimit)
+            newTlsEngine.setHandshakeApplicationProtocolSelector((ex, ps) ->
             {
-                if (tlsRecordInfo.type() == TlsContentType.HANDSHAKE.value())
+                assert newTlsEngine.getHandshakeSession() instanceof ExtendedSSLSession;
+
+                final List<SNIServerName> requestedServerNames = ((ExtendedSSLSession) newTlsEngine.getHandshakeSession())
+                    .getRequestedServerNames();
+                byte[] tlsHostname = !requestedServerNames.isEmpty() ? requestedServerNames.get(0).getEncoded() : null;
+
+                final MessagePredicate filter = (t, b, i, l) ->
                 {
-                    final TlsHandshakeFW tlsHandshake = tlsHandshakeRO.tryWrap(buffer, tlsFragmentOffset, tlsRecordLimit);
+                    final TlsRouteExFW routeExFW = wrapRouteEx.apply(t, b, i, l);
 
-                    if (tlsHandshake != null && tlsHandshake.type() == TlsHandshakeType.CLIENT_HELLO.value())
-                    {
-                        final TlsClientHelloFW tlsClientHello = tlsHandshake.body().get(tlsClientHelloRO::tryWrap);
+                    final String8FW hostname1 = routeExFW.hostname();
+                    final byte[] hostname = new byte[hostname1.length()];
```

Comment:
new object creation, before I start digging into how we can solve this problem just wanted to check if we have a common solution.

Reply:
How about wrapping the `tlsHostname` bytes in an `UnsafeBuffer` and comparing to the `.value()` of `routeEx.hostname()`? (see above)